compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -1173,6 +1173,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
                 true
             }
+            (&ty::Uint(ref exp), &ty::Char) => {
+                let cast_infallible = exp.bit_width().map(|w| w >= 32).unwrap_or(false);
```

Comment:
This is probably better written as

```rust
(&ty::Uint(ref exp), &ty::Char) if exp.bit_width() >= Some(32) => {...}
```

One other thing that is quite unfortunate is the duplication between match arms, but I don't see an easy way to avoid it outside of listing all of the variants. Perhaps that would be better either way, e.g. something like the snippet below doesn't seem all that verbose:

```rust
( ty::Uint(ty::UintTy::U32 | ty::UintTy::U64 | ty::UintTy::U128) 
  | ty::Int(ty::IntTy::I32, ty::IntTy::I64, ty::IntTy::I128), &ty::Char)
```